### ISSUE TITLE
oadnausdjpy1hrun

### ISSUE THREAD

**Issue by kubohiroki-kovac** [kubohiroki-kovac/usdjpy#1]
import configparser

import pandas as pd
import oandapy
import datetime
from datetime import datetime, timedelta
import pytz
import talib as ta

import numpy as np
import matplotlib.pyplot as plt
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import mean_absolute_error
import tensorflow as tf

import oandapyV20
import oandapyV20.endpoints.orders as orders

# 文字列 -> datetime
def iso_to_jp(iso):
    date = None
    try:
        date = datetime.strptime(iso, '%Y-%m-%dT%H:%M:%S.%fZ')
        date = pytz.utc.localize(date).astimezone(pytz.timezone("Asia/Tokyo"))
    except ValueError:
        try:
            date = datetime.strptime(iso, '%Y-%m-%dT%H:%M:%S.%f%z')
            date = date.astimezone(pytz.timezone("Asia/Tokyo"))
        except ValueError:
            pass
    return date

# 設定
config = configparser.ConfigParser()
config.read('./config_practice.txt') 
account_id = config['oanda']['account_id']
api_key = config['oanda']['api_key']

# APIへ接続
oanda = oandapy.API(environment="practice", 
                    access_token=api_key)

api = oandapyV20.API(access_token=api_key, environment="practice")

res_hist_1m_5000 = oanda.get_history(instrument="USD_JPY", granularity="H1", count="5000")
res_hist_1m_5000 = pd.DataFrame(res_hist_1m_5000['candles'])
res_hist_1m_5000=res_hist_1m_5000[:-1]

 # 日付をISOから変換
res_hist_1m_5000['time'] = res_hist_1m_5000['time'].apply(lambda x: iso_to_jp(x))
res_hist_1m_5000['hour'] = res_hist_1m_5000['time'].apply(lambda x: x.hour)
res_hist_1m_5000['weekday'] = res_hist_1m_5000['time'].apply(lambda x: x.weekday())

 # 書き出し用カラムセット
tech_columns=["closeBid",
              "highBid",
              "lowBid",
              "openBid",
              "weekday",
              "hour",
              "volume",
              "RSI",
              "bandmean",
              "bandupper",
              "bandlower",
              "williamR",
              "adx",
              "dip",
              "dim",
              "atr",
              "tma",
              "slowk",
              "slowd",
              "fastk",
              "fastd",
              "ULTOSC",
              "ADXR",
              "APO",
              "aroondown",
              "aroonup",
              "AROONOSC",
              "BOP",
              "CCI",
              "CMO",
              "DX",
              "macd",
              "macdsignal",
              "macdhist",
              "macdext",
              "macdextsignal",
              "macdexthist",
              "macdfix",
              "macdfixsignal",
              "macdfixhist",
              "MFI",
              "MINUS_DI",
              "MINUS_DM",
              "MOM",
              "PLUS_DI",
              "PLUS_DM",
              "PPO",
              "RSI"]

# テクニカル指標データセット
tech = pd.DataFrame()

# ボリンジャーバンド
tech['bandmean'] = res_hist_1m_5000['closeBid'].rolling(window=20).mean()
tech['bandstd'] = res_hist_1m_5000['closeBid'].rolling(window=20).std()
tech['bandupper'] = tech['bandmean'] + (tech['bandstd'] * 2)
tech['bandlower'] = tech['bandmean'] - (tech['bandstd'] * 2)

# wiliam%R
tech['williamR'] = ta.WILLR(res_hist_1m_5000['highBid'], res_hist_1m_5000['lowBid'], res_hist_1m_5000['closeBid'], timeperiod=14)

#　DMI
tech["adx"] = ta.ADX(res_hist_1m_5000['highBid'], res_hist_1m_5000['lowBid'], res_hist_1m_5000['closeBid'], 14)
tech["dip"] = ta.PLUS_DI(res_hist_1m_5000['highBid'], res_hist_1m_5000['lowBid'], res_hist_1m_5000['closeBid'], 14)
tech["dim"] = ta.MINUS_DI(res_hist_1m_5000['highBid'], res_hist_1m_5000['lowBid'], res_hist_1m_5000['closeBid'], 14)

# atr
tech["atr"] = ta.ATR(res_hist_1m_5000['highBid'], res_hist_1m_5000['lowBid'], res_hist_1m_5000['closeBid'], 20)

# tma
tech["tma"]=ta.TRIMA(res_hist_1m_5000['closeBid'], 20)

# STOCH
tech["slowk"], tech["slowd"] = ta.STOCH(res_hist_1m_5000['highBid'], res_hist_1m_5000['lowBid'], res_hist_1m_5000['closeBid'], fastk_period=5, slowk_period=3, slowk_matype=0, slowd_period=3, slowd_matype=0)

# STOCHRSI
tech["fastk"], tech["fastd"] = ta.STOCHRSI(res_hist_1m_5000['closeBid'], timeperiod=14, fastk_period=5, fastd_period=3, fastd_matype=0)

# Ultimate Oscillator
tech["ULTOSC"]=ta.ULTOSC(res_hist_1m_5000['highBid'], res_hist_1m_5000['lowBid'], res_hist_1m_5000['closeBid'], timeperiod1=7, timeperiod2=14, timeperiod3=28)

# Average Directional Movement Index Rating
tech["ADXR"]=ta.ADXR(res_hist_1m_5000['highBid'], res_hist_1m_5000['lowBid'], res_hist_1m_5000['closeBid'], timeperiod=14)

# Absolute Price Oscillator
tech["APO"]=ta.APO(res_hist_1m_5000['closeBid'], fastperiod=12, slowperiod=26, matype=0)

# Aroon
tech["aroondown"], tech["aroonup"] = ta.AROON(res_hist_1m_5000['highBid'], res_hist_1m_5000['lowBid'], timeperiod=14)

# Aroon Oscillator
tech["AROONOSC"]=ta.AROONOSC(res_hist_1m_5000['highBid'], res_hist_1m_5000['lowBid'], timeperiod=14)

# Balance Of Power
tech["BOP"]=ta.BOP(res_hist_1m_5000['openBid'], res_hist_1m_5000['highBid'], res_hist_1m_5000['lowBid'], res_hist_1m_5000['closeBid'])

# Commodity Channel Index
tech["CCI"]=ta.CCI(res_hist_1m_5000['highBid'], res_hist_1m_5000['lowBid'], res_hist_1m_5000['closeBid'], timeperiod=14)

# Chande Momentum Oscillator
tech["CMO"]=ta.CMO(res_hist_1m_5000['closeBid'], timeperiod=14)

# Directional Movement Index
tech["DX"]=ta.DX(res_hist_1m_5000['highBid'], res_hist_1m_5000['lowBid'], res_hist_1m_5000['closeBid'], timeperiod=14)

# macd
tech["macd"],tech["macdsignal"],tech["macdhist"]=ta.MACD(res_hist_1m_5000['closeBid'],fastperiod=12, slowperiod=26, signalperiod=5)

# MACD with controllable MA type
tech["macdext"], tech["macdextsignal"], tech["macdexthist"] = ta.MACDEXT(res_hist_1m_5000['closeBid'], fastperiod=12, fastmatype=0, slowperiod=26, slowmatype=0, signalperiod=9, signalmatype=0)

# Moving Average Convergence/Divergence Fix 12/26
tech["macdfix"], tech["macdfixsignal"], tech["macdfixhist"] = ta.MACDFIX(res_hist_1m_5000['closeBid'], signalperiod=9)

# Money Flow Index
tech["MFI"]=ta.MFI(res_hist_1m_5000['highBid'], res_hist_1m_5000['lowBid'], res_hist_1m_5000['closeBid'], res_hist_1m_5000['volume'], timeperiod=14)

# Minus Directional Indicator
tech["MINUS_DI"]=ta.MINUS_DI(res_hist_1m_5000['highBid'], res_hist_1m_5000['lowBid'], res_hist_1m_5000['closeBid'], timeperiod=14)

# Minus Directional Movement
tech["MINUS_DM"]=ta.MINUS_DM(res_hist_1m_5000['highBid'], res_hist_1m_5000['lowBid'], timeperiod=14)

# Momentum
tech["MOM"]=ta.MOM(res_hist_1m_5000['closeBid'], timeperiod=10)

# Plus Directional Indicator
tech["PLUS_DI"]=ta.PLUS_DI(res_hist_1m_5000['highBid'], res_hist_1m_5000['lowBid'], res_hist_1m_5000['closeBid'], timeperiod=14)

# Plus Directional Movement
tech["PLUS_DM"]=ta.PLUS_DM(res_hist_1m_5000['highBid'], res_hist_1m_5000['lowBid'], timeperiod=14)

# Percentage Price Oscillator
tech["PPO"]=ta.PPO(res_hist_1m_5000['closeBid'], fastperiod=12, slowperiod=26, matype=0)

# Relative Strength Index
tech['RSI']=ta.RSI(res_hist_1m_5000['closeBid'], timeperiod=14)

# テクニカル指標結合
res_hist_1m_5000=pd.concat([res_hist_1m_5000, tech], axis=1)

# NULLデータを含むデータを削除
res_hist_1m_5000=res_hist_1m_5000[40:]

#res_hist_1m_5000.to_csv('usdjpy1h.csv',index=False)

res_hist_1m_5000.to_csv('usdjpy1h.csv',columns=tech_columns,index=False)

df = pd.read_csv("usdjpy1h.csv")

df_shift = df.copy()
df_shift.closeBid = df_shift.closeBid.shift(-1)
#df_shift = df_shift[:-1]

df_2 = df_shift.copy()
#del df_2['time']

# データセットの行数と列数を格納
n = df_2.shape[0]
p = df_2.shape[1]
 

train_start = 0
train_end = int(np.floor(0.8*n))
test_start = train_end + 1
test_end = n
data_train = df_2.loc[np.arange(train_start, train_end), :]
data_test = df_2.loc[np.arange(test_start, test_end), :]

# データの正規化
scaler = MinMaxScaler(feature_range=(-1, 1))
scaler.fit(data_train)
data_train_norm = scaler.transform(data_train)
data_test_norm = scaler.transform(data_test)

# 特徴量とターゲットへ切り分け
X_train = data_train_norm[:, 1:]
y_train = data_train_norm[:, 0]
X_test = data_test_norm[:, 1:]
y_test = data_test_norm[:, 0]

# 正規化から通常の値へ戻す
y_test = y_test.reshape(991, 1)
test_inv = np.concatenate((y_test, X_test), axis=1)
test_inv = scaler.inverse_transform(test_inv)

# 訓練データの特徴量の数を取得
n_stocks = X_train.shape[1]
 
# ニューロンの数を設定
#n_neurons_1 = 256
#n_neurons_2 = 128

n_neurons_1 = 2048
n_neurons_2 = 1024

# セッションの開始
net = tf.InteractiveSession()
 
# プレースホルダーの作成
X = tf.placeholder(dtype=tf.float32, shape=[None, n_stocks])
Y = tf.placeholder(dtype=tf.float32, shape=[None])
 
# 初期化
sigma = 1
weight_initializer = tf.variance_scaling_initializer(mode="fan_avg", distribution="uniform", scale=sigma)
bias_initializer = tf.zeros_initializer()

# Hidden weights
W_hidden_1 = tf.Variable(weight_initializer([n_stocks, n_neurons_1]))
bias_hidden_1 = tf.Variable(bias_initializer([n_neurons_1]))
W_hidden_2 = tf.Variable(weight_initializer([n_neurons_1, n_neurons_2]))
bias_hidden_2 = tf.Variable(bias_initializer([n_neurons_2]))

# 出力の重み
W_out = tf.Variable(weight_initializer([n_neurons_2, 1]))
bias_out = tf.Variable(bias_initializer([1]))

# 隠れ層の設定（ReLU＝活性化関数）
hidden_1 = tf.nn.relu(tf.add(tf.matmul(X, W_hidden_1), bias_hidden_1))
hidden_2 = tf.nn.relu(tf.add(tf.matmul(hidden_1, W_hidden_2), bias_hidden_2))
 
# 出力層の設定
out = tf.transpose(tf.add(tf.matmul(hidden_2, W_out), bias_out))

# コスト関数
mse = tf.reduce_mean(tf.squared_difference(out, Y))

# 最適化関数
opt = tf.train.AdamOptimizer().minimize(mse)

# 初期化
net.run(tf.global_variables_initializer())

# コスト関数
mse = tf.reduce_mean(tf.squared_difference(out, Y))
 
# 最適化関数
opt = tf.train.AdamOptimizer().minimize(mse)
 
# 初期化
net.run(tf.global_variables_initializer())

# ニューラルネットワークの設定
batch_size = 256
mse_train = []
mse_test = []
 
# 訓練開始！500回の反復処理
epochs = 500
for e in range(epochs):
    net.run(opt, feed_dict={X: X_train, Y: y_train})

# テストデータで予測
pred_test = net.run(out, feed_dict={X: X_test})
 

# 予測値をテストデータに戻そう（値も正規化からインバース）
pred_test = np.concatenate((pred_test.T, X_test), axis=1)
pred_inv = scaler.inverse_transform(pred_test)

print(data_test.values[990][0])
 
# テストデータの最後のデータ（正規化を戻した後）
print(test_inv[990][0])
 
# モデルが予測したデータ
print(pred_inv[990][0])

# MAEの計算
mae_test = mean_absolute_error(test_inv[:-1], pred_inv[:-1])
print(mae_test)

res_mlt = oanda.get_prices(instruments="USD_JPY")

print(res_mlt['prices'][0]["bid"])

if pred_inv[990][0]+0.03<res_mlt['prices'][0]["bid"]:
    data = {"order": {
                "instrument": "USD_JPY",
                "units": "-1000",
                "type": "MARKET",
                "positionFill": "DEFAULT",
                "takeProfitOnFill": {
                    "timeInForce": "GTC",
                    "price": str('{0:.3f}'.format(pred_inv[990][0],3))
                },
                "stopLossOnFill": {
                    "timeInForce": "GTC",
                    "price": str(res_mlt['prices'][0]["bid"]+0.16)
                }
                
              }
            }
    
    # 成行注文
    r = orders.OrderCreate(account_id, data=data)
    api.request(r)
    
    

elif pred_inv[990][0]-0.03>res_mlt['prices'][0]["bid"]:
    data = {"order": {
                "instrument": "USD_JPY",
                "units": "+1000",
                "type": "MARKET",
                "positionFill": "DEFAULT",
                "takeProfitOnFill": {
                    "timeInForce": "GTC",
                    "price": str('{0:.3f}'.format(pred_inv[990][0],3))
                },
                "stopLossOnFill": {
                    "timeInForce": "GTC",
                    "price": str(res_mlt['prices'][0]["bid"]-0.16)
                }
              }
            }
    # 成行注文
    r = orders.OrderCreate(account_id, data=data)
    api.request(r)
